clustering.Rmd

```diff
@@ -541,57 +662,65 @@ penguin_clust_ks <- tibble(k = 1:9)
+data structures for data frame columns. 
+To make this work, 
+we have to put each model object in a list using the `list` function.
+We demonstrate how to do this below:
+
+```{r}
+penguin_clust_ks <- tibble(k = 1:9) |>
+  rowwise() |>
+  mutate(penguin_clusts = list(kmeans(standardized_data, k)))
 ```
 
-The anonymous function in the above call is `function(k) kmeans(scaled_data, k)`.
-This function takes a single argument (`k`) and evaluates `kmeans(scaled_data, k)`.
-Since `k` is the *first* (and only) argument to the function, we can use `map` just
-like we did before! The rest of the call above does just that -- it passes each row of
-`penguin_clust_ks` to our anonymous function.
+If we take a look at our data frame `penguin_clust_ks` now, 
+we see that it has two columns: one with the value for K, 
+and the other holding the clustering model object in a list column.
 
-Below, we execute this `map` call inside of a `mutate` call on the 
-`penguin_clust_ks` data frame and get a list column that contains a K-means
-clustering object for each value of K we had:
-\index{mutate}
-
-```{r 10-choose-k-part2}
-penguin_clust_ks <- tibble(k = 1:9) |>
-  mutate(penguin_clusts = map(k, function(ks) kmeans(scaled_data, ks)))
+```{r}
 penguin_clust_ks
 ```
 
-Next, we use `map` again to apply `glance` \index{glance} to each of the K-means
-clustering objects to get the clustering statistics (including WSSD). The output
-of glance is a data frame, and so we get another list column. This results in a 
-complex data frame with 3 columns, one for K, one for the 
+> If we wanted to get one of the clusterings out 
+> of the list column in the data frame,
+> we can use our familiar friends `slice` and `pull`.
+> However, before we do this, we need to turn `rowwise`.
+> To do that, we use `ungroup` 
+> (predictably, this can also turn `group_by` off as well).
+> 
+> ```{r}
+> penguin_clust_ks |>
+>   ungroup() |>
+>   slice(1) |>
+>   pull(penguin_clusts)
+> ```
+
+Next, we use `mutate` again to apply `glance` \index{glance} 
+to each of the K-means clustering objects to get the clustering statistics 
+(including WSSD). 
+The output of glance is a data frame, 
+and so we need to create another list column (using `list`) for this to work. 
+This results in a complex data frame with 3 columns, one for K, one for the 
 K-means clustering objects, and one for the clustering statistics:
 
 ```{r 10-choose-k-part3}
 penguin_clust_ks <- tibble(k = 1:9) |>
-  mutate(
-    penguin_clusts = map(k, ~ kmeans(scaled_data, .x)),
-    glanced = map(penguin_clusts, glance)
-  )
+  rowwise() |>
+  mutate(penguin_clusts = list(kmeans(standardized_data, k)),
+         glanced = list(glance(penguin_clusts)))
+
 penguin_clust_ks
 ```
 
-Finally we extract the total WSSD from the `glanced` column. Given that each 
-item in this column is a data frame, we will need to use the `unnest` function 
+Finally we extract the total WSSD from the `glanced` column. 
```

Comment:
`glanced` is not a function, right? should rephrase this just to use the standard `glance` function name

Reply:
No, the column name is `glanced`. I will make that more clear.